Documentation/llilc-at-six-months.md

```diff
@@ -0,0 +1,456 @@
+MSIL and translating them into vector LLVM IR.
+
+For example, when jitting code like
+
+```C#
+  using Point = System.Numerics.Vector2;
+
+  Point a = new Point(10, 50);
+  Point b = new Point(10, 10);
+  Point c = a * b;
+```
+
+LLILC produces the following LLVM IR:
+
+```LLVM
+  %1 = addrspacecast <2 x float>* %loc0 to <2 x float> addrspace(1)*, !dbg !12
+  store <2 x float> <float 1.000000e+01, float 5.000000e+01>, 
+        <2 x float> addrspace(1)* %1, !dbg !18
+  %2 = addrspacecast <2 x float>* %loc1 to <2 x float> addrspace(1)*, !dbg !19
+  store <2 x float> <float 1.000000e+01, float 1.000000e+01>, 
+        <2 x float> addrspace(1)* %2, !dbg !20
+  %3 = load <2 x float>, <2 x float>* %loc0, align 8, !dbg !21
+  %4 = load <2 x float>, <2 x float>* %loc1, align 8, !dbg !21
+  %5 = fmul <2 x float> %3, %4, !dbg !22
+  store <2 x float> %5, <2 x float>* %loc2, !dbg !22
+  %6 = addrspacecast <2 x float>* %loc2 to <2 x float> addrspace(1)*, !dbg !23
+```
+
+and in turn this produces the following assembly code:
+
+```Assembly
+  mov         rax,7F909DA0510h  
+  mov         rax,qword ptr [rax]  
+  mov         qword ptr [rsp+38h],rax   
+  mov         rax,7F909DA0520h  
+  movaps      xmm0,xmmword ptr [rax]      
+  movlps      qword ptr [rsp+30h],xmm0  
+  movq        xmm1,mmword ptr [rsp+38h]  
+  mulps       xmm1,xmm0  
+  movlps      qword ptr [rsp+28h],xmm1
+```
+
+## Extending LLVM: GC
+
+In the [GC document](llilc-gc.md) we gave an overview of the challenges we face in
+supporting the CoreCLR�s GC. Much of our work to date has been facilitated by
+enabling a conservative GC mode in the CoreCLR while we wait for our precise GC
+implementation to mature. 
+
+The CoreCLR�s GC is a precise, relocating collector. This means that at places
+where GC can occur (safepoints), all heap references must be accurately
+described to the GC, and all of them may change values at a GC.
+
+For precise GC, LLILC has built upon the statepoint representation for GC
+safepoints. Safepoints are modelled as calls, and the possible modifications to
+GC references are made explicit in the IR as extra arguments to (and �return 
+values� from) the statepoint wrapper for the calls. This information is later
+used to explicitly spill these references to memory, and then report those
+locations via LLVM-created per-safepoint stackmaps. 
+
+LLILC parses the stackmaps to produce the �GC Info� format that the CoreCLR
+expects. The CLR format required some new data we couldn�t get from the
+original stackmaps, so we worked along with the community to create �V2�
+enhanced stackmaps.  So far, this has worked well. 
+
+The CoreCLR�s GC mode (conservative or precise) is set when the runtime
+initializes. By leveraging the alt-jit mechanism we can enable precise GC and
+run tests end-to-end, allowing LLILC to bail out of methods where it can�t
+generate correct GC info. We�re now able to jit the majority of methods using
+LLILC with precise GC enabled in this fashion.
+
```

Comment:
We've also extended the Statepoint rewriting algorithms to deal with managed pointers. That is, (a) Extend the liveness algorithm to track integer<->pointer conversions, and (b) optimize the reporting algorithm to not report pointers known to be out of the managed heap (ex: stack locations).
[This change is in the MS branch].